### PR TITLE
Enable two slot fifo for all crsf ext frames

### PR DESCRIPTION
**Pull request by schugabe**
Change dropping behaviour: overwrite oldest message in fifo that is not sent yet

TBD: test with different FC projects if this change breaks anything

More generic PR based on https://github.com/pmattila/ExpressLRS/commit/46d1a88dc1028d8df2a4ef3cf23c2ea80c63aac0

### CODE REVIEW

**Comment by JyeSmith** [comment]
> TBD: test with different FC projects if this change breaks anything

Looking all ok with Ardupilot.

**Comment by wvarty** [comment]
Tested with ardu using a ranger and a BAYCK nano RX.
LUA menu for ardu loads in other devices, and params load and can be changed successfully.
Model telm coming through fine.
Seems to be working.

**Review by wvarty — APPROVED** [review]

File: src/lib/Telemetry/telemetry.cpp [hunk 1]
```diff
@@ -144,6 +144,7 @@ void Telemetry::ResetState()
     telemetry_state = TELEMETRY_IDLE;
     currentTelemetryByte = 0;
     currentPayloadIndex = 0;
+    currentQueueIndex = 0;
```
Comment (CapnBry) [inline src/lib/Telemetry/telemetry.cpp:147]: This whole thing's a queue! Please use less generic term for this feature, suggest maybe `twoslotLastQueueIndex`?

**Review by CapnBry — CHANGES_REQUESTED** [review]
Thanks again for taking the initiative, as I think this drop-oldest 2-slot method is better than our existing drop-newest.

Tested with Betaflight CMS and MSP-based lua. As expected, performance is subjectively the same. Also tested ELRS RX (CRSF-based) config just to be sure there was no unintended interaction and there was not, so all good there too. Still a little miffed that the original author couldn't be bothered to do their own testing, but big thanks to Jye and Wez for doing the work!

I couldn't think of what to test with iNav with this, is there any MSP-based iNav lua? All I am aware of is the telemetry-only one.

**Comment by JyeSmith** [comment]
@rotorflight Are you able to test?

**Comment by rotorflight** [comment]
I'll test it. 

Though, this code was not meant to be used with all frame types, as it has a minor side effect. Occasionally it is reordering the frames, i.e. they don't get send out in the same order they were received. That's ok for our telemetry, but it may well break other protocols.

In fact, I concluded in my later testing that the code currently used with MSP messages is the best compromise. It does not reorder frames. And it is tail-drop, which is probably the most common way.

**Comment by rotorflight** [comment]
I kept this code running overnight without any issues with our custom telemetry. I also tested MSP functionality in Rotorflight. Everything seems to be working ok.

I also compared this to the code I was suggesting earlier (the MSP queue method expanded to all custom frames)
https://github.com/pmattila/ExpressLRS/commit/a58e7a1cf8cfc4d478acd0b764050cbe7aec5b16

For our purposes this works exactly the same, I have not seen any issues with either approach.

I am in favour of the latter, as it does not change the behaviour with MSP frames, and the tail drop queue is probably more expected/common than head-drop. 

But Rotorflight telemetry works with either approach, so either way would be fine for RF.

File: src/lib/Telemetry/telemetry.cpp (hunk 1 above)
Comment (schugabe) [reply src/lib/Telemetry/telemetry.cpp]: sounds good

**Review by CapnBry — APPROVED** [review]
I did consider that this change would make it so incoming packets A B C D would come out A D instead of A B (as it would be currently), but nothing I've seen is tolerant of either and bombs on both. There is a chance that A+B is a complete frame and could be used by the recipient, but there's practically as good a chance it would come as C D and that C+D is a complete usable frame. I consider these to be roughly equal, at least with current protocols I am aware of.

However, Rotorflight is correct, I believe this code also has a higher chance of delivering packets out of order as our existing code due to the sender always progressing through the slots in order, and them being filled in an alternating fashion in the new code if both slots are full. I think this could be resolved by always using the last slot and moving an existing packet up a slot when the older one is replaced.

Ultimately, I think there is slightly more benefit to being able to replace stale queued packets with fresher ones compared to the two disadvantages above. The first feels like a wash, but the second could definitely be an issue when older telemetry is received after newer telemetry.

I'm fine with doing this or not doing this or doing this a different way, so long as the behavior doesn't become more specific to certain packet types.

**Pull request merged**